[1-3단계 - 방탈출 예약 관리] 벨로(방예혁) 미션 제출합니다.

YehyeokBang opened the pull request:
안녕하세요! 제로!
리뷰어로 만나게되어 영광입니다! 👍 👍 
잘 부탁드립니다!

<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->
## 체크 리스트
- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [x] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
  - [x] [벨로의 프롤로그](https://prolog.techcourse.co.kr/studylogs/4221) 

## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?
### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [ ] 3 (보통)
- [x] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.
<!-- 이유 작성 -->
객체지향 생활체조 원칙을 지키려고 하는데, 레벨1과는 뭔가 다름을 느낀 것 같습니다. 😢
(스프링 부트로 구현하는 것에 초점이 맞춰진 미션이라서 그랬나?라는 생각이 들었습니다.)
그래도 어떤 책임을 가져야 하는지 고민하고 구현하려고 했습니다.

## 웹 개발 경험 체크
🌱 레벨2 시작 이전에, 웹 애플리케이션 개발 경험이 있나요?
- [ ] 아니오. 전혀 경험이 없습니다.
- [ ] 예. 예제 코드를 따라하거나 클론 코딩을 해봤습니다. (ex. 강의를 보고 따라함)
- [ ] 예. 웹 애플리케이션을 만들었지만 로컬에서만 실행해봤습니다.
- [x] 예. 웹 애플리케이션을 만들고 배포해봤습니다. (ex. 클라우드 혹은 리모트 서버에 배포)

## 어떤 부분에 집중하여 리뷰해야 할까요?
<!-- 리뷰어가 효과적으로 피드백할 수 있도록 중점적으로 리뷰 받고 싶은 내용을 *요약* 형태로 작성해 주세요.
리뷰해야 할 포인트를 요약해 강조해 주시면, 리뷰어가 코드 전체를 이 부분에 집중해 리뷰할 수 있습니다.
반면 특정 코드부분에 대한 피드백이 필요하다면, 이 곳에 적기 보다 해당 코드를 선택하고 코멘트를 남기는 것을 권장합니다. -->

코드에 더 자세한 코멘트를 남기겠습니다!

### Repository 구현

구현 중 컨트롤러에서 예약 정보를 관리하는 책임까지 가지고 있는 것 같아서.
가장 범용적으로 사용되는 `Repository`라는 이름을 사용하여 이를 분리했습니다.
이를 인터페이스로 두고 필요한 구현체를 만들어서 사용하였습니다.

[인터페이스 분리](https://github.com/woowacourse/spring-roomescape-admin/pull/265#discussion_r2048481515)

### 예외 핸들링

비즈니스 적으로 불가능한 일이 일어날 때 예외를 던지도록 구현했습니다.
덕분에 Controller에서 try-catch 블럭을 사용하여 예외가 발생하지 않으면 200,
발생하면 400을 반환하도록 구현하게 되었습니다. 문제라고 인식하고 예외 핸들러를 활용하여 해결했습니다.
추가로 새로운 예외를 구현하여 사용했습니다.

[겪은 문제](https://github.com/woowacourse/spring-roomescape-admin/pull/265#discussion_r2048475170)
[구현한 예외](https://github.com/woowacourse/spring-roomescape-admin/pull/265#discussion_r2048460482)

### 스프링 부트 테스트

[관련 코멘트](https://github.com/woowacourse/spring-roomescape-admin/pull/265#discussion_r2048504604)

### 📌 코드리뷰 질문 그라운드룰 (지우지 마세요)
✅ 많은 사람과 공유하고 토론해보는 것이 더 좋은 질문에는, 리뷰어가 💬말풍선 이모지를 남겨 토론을 추천할 수 있어요.  
✅ 토론을 추천받은 질문은 `#be-7기-잡담-질답`슬랙채널이나 캠퍼스에서 자유롭게 토론을 이어가보세요!  
✅ 코드에 대한 맥락 없이 리뷰어의 주관적인 의견을 묻거나, 단순히 현업에서는 어떻게 하는지를 묻는 질문은 코드 리뷰 본문에는 어울리지 않아요.  
코드 리뷰는 변경된 코드의 목적, 의도, 품질을 중심으로 논의하는 자리예요.  
✅ 코드와 관련된 질문이 있다면 PR 본문에 적기보다는, 해당 코드를 선택해 코멘트를 남겨주세요.

YehyeokBang committed:
docs: 구현할 기능 목록 작성

Co-authored-by: saera-yook <yook.saera@gmail.com>

YehyeokBang committed:
style: 코드 정렬

Co-authored-by: saera-yook <yook.saera@gmail.com>

YehyeokBang committed:
feat: 관리자 홈 화면 구현

Co-authored-by: saera-yook <yook.saera@gmail.com>

YehyeokBang committed:
feat: 예약 모델 구현

Co-authored-by: saera-yook <yook.saera@gmail.com>

YehyeokBang committed:
feat: 예약 조회 화면 구현

Co-authored-by: saera-yook <yook.saera@gmail.com>

YehyeokBang committed:
feat: 예약 추가 기능 구현

Co-authored-by: saera-yook <yook.saera@gmail.com>

YehyeokBang committed:
feat: 예약 취소 기능 구현

Co-authored-by: saera-yook <yook.saera@gmail.com>

YehyeokBang committed:
fix: 예약 화면 이동 버튼의 링크 경로 수정

Co-authored-by: saera-yook <yook.saera@gmail.com>

YehyeokBang committed:
refactor: 모델 대신 DTO를 반환하도록 변경

Co-authored-by: saera-yook <yook.saera@gmail.com>

YehyeokBang committed:
feat: 예약 생성 시 날짜 및 시간 검증 구현

- 과거 일시로 예약을 생성할 수 없습니다.

Co-authored-by: saera-yook <yook.saera@gmail.com>

YehyeokBang committed:
feat: 예약 생성 시 필드 null 검증 구현

Co-authored-by: saera-yook <yook.saera@gmail.com>

YehyeokBang committed:
refactor: 패키지 분리

Co-authored-by: saera-yook <yook.saera@gmail.com>

YehyeokBang committed:
refactor: 예외 처리 방식 변경

- 기존 컨트롤러에서 try-catch 블럭으로 처리하던 방식에서 핸들러 방식으로 변경합니다.

Co-authored-by: saera-yook <yook.saera@gmail.com>

YehyeokBang committed:
refactor: 컨트롤러 유형에 맞게 클래스명 변경

- 템플릿 반환 여부에 따라 구분합니다.

Co-authored-by: saera-yook <yook.saera@gmail.com>

YehyeokBang committed:
refactor: 예약 관리 책임 분리

- 기존 컨트롤러에서 책임지던 예약 관리를 저장소로 분리합니다.

Co-authored-by: saera-yook <yook.saera@gmail.com>

YehyeokBang committed:
docs: 애플리케이션 사용 방법 및 예시 작성

Co-authored-by: saera-yook <yook.saera@gmail.com>

YehyeokBang commented on src/main/java/roomescape/exception/DomainException.java:
```diff
@@ -0,0 +1,8 @@
+package roomescape.exception;
+
+public class DomainException extends IllegalArgumentException {
+
+    public DomainException(String message) {
+        super(message);
+    }
+}
```
페어와 함께 도메인과 모델에 대해 아래와 같이 정리하고,
이 정의를 바탕으로 `DomainException`이라는 이름을 사용했습니다.

### 도메인과 모델

페어와 함께 합의했던 내용입니다.

`도메인`은 우리가 해결하려는(더 나아지게 만들려는) 영역을 의미합니다.
우리가 지금까지 진행한 미션에서는 블랙잭이나 장기 게임 등이 도메인이라고 생각합니다.
지금은 방탈출 예약 관리가 도메인이 될 것 같습니다.

`모델`은 도메인에서 실체 혹은 개념적으로 존재할 수 있는 모든 것들을 의미합니다.
예를 들어, 장기에서는 각 기물, 보드판, 좌표 등이 모델이라고 생각합니다.

### DomainException 이름

`DomainException`은 단순한 기술적 오류(예: NullPointerException 등)가 아니라,
도메인 안에서의 규칙을 위반했을 때 발생하는 예외를 의미합니다.

예를 들어, 과거의 날짜로 예약을 만들려고 하거나, 이름 없이 예약을 하려는 경우에 발생합니다.
이런 경우는 모두 예약이라는 도메인 안에서 정의된 규칙을 어긴 상황입니다.
따라서 이런 예외들을 `DomainException`으로 표현하는 것이 괜찮겠다고 생각했습니다.

그러나, 꽤 추상적인 이름이라고 느끼기도 했습니다.
더 세분화를 해야하는지(ReservationException 등) 고민을 했으나,
지금 구조에서는 오히려 더 복잡해질 것 같아서 보류했습니다.

제로의 조언이나 생각이 궁금해요! 😄

asebn1 replied:
DomainException이 너무 포괄적인 의미인 것 같아요
네이밍만 보아서는 어떤 의미를 담는지 알 수 없네요 ㅠ
요건 추후 개선사항으로 체크해놓으면 좋을 것 같아요~

YehyeokBang replied:
넵 더 명확한 이름을 고민해보겠습니다!
감사합니다! 👍 

---

[refactor: 예외 클래스명 변경](https://github.com/woowacourse/spring-roomescape-admin/pull/265/commits/aea05cc9525f58df592677b03302a8040eaaf6e8)

YehyeokBang commented on src/main/java/roomescape/exception/DomainExceptionHandler.java:
```diff
@@ -0,0 +1,16 @@
+import org.springframework.web.bind.annotation.ControllerAdvice;
+import org.springframework.web.bind.annotation.ExceptionHandler;
+
+@ControllerAdvice
+public final class DomainExceptionHandler {
+
+    @ExceptionHandler(DomainException.class)
+    public ResponseEntity<String> handleDomainException(DomainException exception) {
+        return ResponseEntity.status(HttpStatus.BAD_REQUEST)
+                .body(exception.getMessage());
+    }
+}
```
### 겪은 문제

```java
// 겪었던 순간의 코드라서 지금과는 다를 수 있습니다.
@PostMapping("/reservations")
public ResponseEntity<?> reserve(@RequestBody ReservationRequestDto dto) {
    try {
        // 예를 들어 과거의 시간으로 예약을 생성하려는 경우 예외가 발생합니다.
        Reservation reservation = dto.toEntity(id.getAndIncrement());
        reservations.add(reservation);
        return ResponseEntity.ok(ReservationResponseDto.from(reservation));
    } catch (DomainException e) {
        return ResponseEntity.badRequest().body(e.getMessage());
    }
}
```

- Controller가 도메인 예외를 직접 알고, 처리까지 해줘야 한다.
  예외가 어디서 발생하는지를 Controller가 알아야 하고,
  이를 처리하는 로직까지 담당하다 보니 Controller의 책임이 커진다.

- try-catch로 인해 코드 가독성이 떨어진다.
  비즈니스 로직과 에러 처리 로직이 섞여, 핵심 흐름을 파악하기 어렵다.

- `ResponseEntity<?>` 사용이 어색하게 느껴졌다.
    성공과 실패를 모두 포괄하려다 보니, 반환 타입이 명확하지 않다는 점이 마음에 걸린다.

이런 문제를 경험했습니다.

### ExceptionHandler

문제를 해결하기 위해 페어와 `@ExceptionHandler`를 학습하고 적용했습니다.

가장 크게 눈에 보이는 장점은 Controller의 책임이 줄어든 것이라고 생각합니다.
예외 처리에 대한 걱정은 별도의 `DomainExceptionHandler`가 맡습니다.
예전에는 비즈니스 로직의 예외 타입이 바뀌면 해당 catch문도 바꿔야 했지만,
지금은 예외 처리기가 이를 맡아주므로 Controller의 변경을 최소화할 수 있을 것 같습니다.

또한, try-catch 블럭이 사라지고, ResponseEntity를 상황마다 구분할 필요도 없어졌기 때문에
코드가 더 읽기 쉬워진 것 같습니다.

asebn1 replied:
좋습니다 ㅎㅎ
정리까지 잘 해주셨네요! 👍

YehyeokBang commented on src/main/java/roomescape/repository/ReservationRepository.java:
```diff
@@ -0,0 +1,13 @@
+import java.util.List;
+import roomescape.model.Reservation;
+
+public interface ReservationRepository {
+
+    List<Reservation> findAll();
+
+    Reservation add(Reservation reservation);
+
+    void removeById(int id);
+}
```
저장소를 쓰는 클라이언트는 `예약을 어떻게 관리할지`에는 관심을 가지지 않아도 되도록 하고 싶었습니다.
단지, `예약을 관리해줄 무언가`가 필요할 뿐이다! 가 의도였습니다.

그래서 `ReservationRepository`라는 인터페이스를 만들어,
예약 저장 방식에 대한 구체적인 구현은 숨기고 역할만 정의했습니다.

현재는 실행 중인 애플리케이션 내에서 예약 정보를 메모리에 저장하는
`MemoryReservationRepository`를 구현체로 사용하고 있습니다.

이렇게 인터페이스를 두면, 나중에 저장소를 파일이나 DB로 바꾸더라도
변경하기 비교적 쉽지 않을까? 라는 기대로 구현했습니다. 😄

YehyeokBang commented on src/main/java/roomescape/model/Reservation.java:
```diff
@@ -0,0 +1,75 @@
+import java.time.LocalTime;
+import roomescape.exception.DomainException;
+
+public final class Reservation {
+
+    private final Integer id;
+    private final String name;
+    private final LocalDate date;
+    private final LocalTime time;
```
현재 구현에서는 예약 저장소가 id를 생성하고 할당해줍니다.
따라서 Reservation 객체를 처음 만들 때는 id를 전달하지 않습니다.

이때 id의 타입을 기본형 int로 선언하면,
전달하지 않아도 자동으로 0으로 초기화되기 때문에
아직 할당되지 않은 상태인지 아닌지를 구분하기 어렵다고 생각했습니다.

그래서 null을 통해 `아직 없음`을 표현할 수 있도록,
Wrapper 클래스인 Integer를 사용하게 되었습니다.

다만, 자바에서 null은 단순히 "초기화되지 않았다"는 의미 외에도
"존재하지 않음" 등 여러 상황을 의미할 수 있다고 생각합니다.
그래서 오히려 혼란을 줄 수도 있지 않을까? 라는 고민이 들었습니다.

제로의 생각이나 조언이 궁금합니다!! 😄

asebn1 replied:
ID의 존재 여부(null vs 값 있음)로 저장 전/후 상태를 구분하는 것이
도메인 모델에서 객체의 생명주기를 추적하는 일반적인 방법 중 하나라고 생각되어서 좋다고 생각해요

asebn1 replied:
추가로 id는 보통 Long 타입으로 많이 사용하는 것 같아요
지금도 나쁘진 않습니다.

YehyeokBang replied:
> ID의 존재 여부(null vs 값 있음)로 저장 전/후 상태를 구분하는 것이
도메인 모델에서 객체의 생명주기를 추적하는 일반적인 방법 중 하나라고 생각되어서 좋다고 생각해요

좋은 리뷰 감사합니다! 
말씀해주신 id 존재 여부로 객체의 생명주기를 구분하는 방식이 도메인 모델에서 일반적으로 쓰인다는 점이 제 마음을 진정시켜줬어요 🥹

그동안 id를 단순히 객체를 유일하게 식별하는 용도로만 생각했었는데,
생명주기 관점에서도 중요한 역할을 한다는 걸 알게 되어 많이 배웠습니다! 감사합니다 👍

YehyeokBang replied:
> 추가로 id는 보통 Long 타입으로 많이 사용하는 것 같아요

해당 프로젝트에서 현실적으로 21억 개가 넘는 row가 생길 가능성은 극히 드물다고 생각해서, 처음엔 Integer를 사용했습니다.

하지만 왜 Long을 보통 사용하는지 궁금해서 찾아봤고, 말씀하신 것처럼 안정성을 위해 Long을 기본값처럼 사용하는 문화가 많은 것 같다는 걸 알게 되었습니다. (진짜 21억 개가 넘어가는 경우까지 염두에 두는 관점)

그리고 여러 협업 프로젝트나 프레임워크 예제에서 Long을 일관되게 사용하는 흐름도 존재해서, 유지보수나 협업 측면에서도 무난한 선택이란 생각이 들었습니다.

실제로 [Spring Data JPA Docs](https://docs.spring.io/spring-data/jpa/reference/jpa/getting-started.html#redis.hello-world)의 예제에서도 Long을 사용하는 걸 확인할 수 있었어요! 😄

아직 데이터베이스를 사용하지 않았기 때문에, DB의 정수 타입까지 고려하진 않았지만,
이후 단계를 구현하면서 id 타입도 꼭 다시 고민해보겠습니다. 좋은 리뷰 감사해요! 👍 👍

YehyeokBang commented on src/test/java/roomescape/MissionStepTest.java:
```diff
@@ -9,12 +17,61 @@
 @DirtiesContext(classMode = DirtiesContext.ClassMode.BEFORE_EACH_TEST_METHOD)
 public class MissionStepTest {
 
+    @DisplayName("1단계 - 관리자 홈 화면 응답 성공")
     @Test
-    void 일단계() {
+    void step1() {
         RestAssured.given().log().all()
-                .when().get("/")
+                .when().get("/admin")
                 .then().log().all()
                 .statusCode(200);
     }
```
미션에서 주어진 단계별 테스트 코드만 따로 모아두었습니다! 😄 

API를 테스트하는 코드를 작성해보면서 생긴 궁금한 점이 있습니다!

Spring Boot가 실행된 상태에서 수행되는 테스트가 많아질수록, (`@SpringBootTest)
전체 테스트 시간이 선형적으로 늘어나는 것 같다고 느꼈습니다.
이전 자바로만 수행하던 미션의 테스트 코드 실행 속도와는 체감될 정도로 차이나는 것 같아요.

실제 업무에서도 RestAssured 같은 도구로 API를 통합적으로 테스트하는 경우가 꽤 있을 것 같은데,
이런 테스트가 많아지면 테스트 속도가 느려지는 문제를 해결하기 위해 어떻게 접근하시나요?

이와 관련하여 제로의 경험이 궁금합니다! 😄 

### 그냥 해본 생각
- 테스트를 동시에 여러 개를 수행할 수 있나?
- 스프링을 킬 때 오래 걸리는 것 같은데, 스프링 켜두고 재활용(?)
  - 지금 테스트는 매번 스프링 부트를 실행시키고 API 1개 호출인 것 같습니다.
- 이외에도 오래 걸리는 부분을 분리? 가짜 객체 주입?

asebn1 replied:
SpringBootTest는 실제로 앱을 띄우기 때문에 많이 느려집니다.
대부분 MockMvc나 단위 테스트로 구현하고
RestAssured는 통합 테스트가 필요한 부분에서만 사용하는 것 같아요 :)

YehyeokBang replied:
경험 공유 감사합니다! 👍

YehyeokBang reviewed (commented):
구현 의도 또는 궁금한 점을 코드에 코멘트로 남겨두었습니다.

잘 부탁드립니다! 😄

asebn1 commented on README.md:
```diff
@@ -0,0 +1,43 @@
+
+방탈출 예약을 관리할 수 있는 웹 애플리케이션을 구현합니다.
+
+## 구현할 기능 목록
+
+- [X] 관리자 홈 화면
+    - [X] `GET - /admin` 요청 시 관리자 홈 화면을 응답한다.
+- [X] 예약 조회
+    - [X] `GET - /admin/reservation` 요청 시 예약 관리 화면을 응답한다.
+        - [X] 예약 목록 조회 API를 구현한다.
+            - [X] `GET - /reservations` 요청 시 예약 목록을 반환한다.
+- [X] 예약 추가
+    - [X] 예약 추가 API를 구현한다.
+        - [X] `POST - /reservations` 요청 시 예약을 추가한다.
+        - [X] 과거 일시로 예약할 수 없다. 
+- [X] 예약 취소
+    - [X] 예약 삭제 API를 구현한다.
+        - [X] `DELETE - /reservations/{id}` 요청 시 예약을 취소한다.
+
```
꼼꼼하게 작성해 주셨네요 👍

YehyeokBang replied:
감사합니다! 😄

asebn1 commented on src/main/java/roomescape/controller/ReservationApiController.java:
```diff
@@ -0,0 +1,43 @@
+    }
+
+    @PostMapping
+    public ReservationResponseDto reserve(@RequestBody ReservationRequestDto reservationRequestDto) {
```
만약 입력값 유효성 검증이 추가되어야 한다면 어디에 추가되어야 할까요?

YehyeokBang replied:
추가되는 입력값 유효성 검증을 생각해봤을 때,
크게 두 가지(형식 검증, 비즈니스 규칙)로 나뉠 것 같습니다.

## 형식 검증

입력값이 기술적으로 유효한가를 판단합니다.
주로 컨트롤러 앞단에서 처리하는 것이 효율적이라고 생각합니다.

- null인가? 빈 문자열인가? 숫자인가? 등

## 비즈니스 규칙

입력값이 도메인 개념상 유효한가를 판단합니다.
모델 내부에서 비즈니스 규칙으로 다루는 것이 적절하다고 생각합니다.

- 예약 시간이 현재 시간보다 이전일 수는 없다. 출석 가능 시간이 아니라면 출석 체크를 진행할 수 없다. 금액은 음수가 될 수 없다. 등

### 그런데…

저는 모델 내에서 null을 검증하고 있습니다.
이건 단순히 사용자가 null을 넘겼다는 걸 검증하기보다,
`개발자가 잘못된 객체를 만들지 못하게 설계하고 싶었다`는 의도가 컸습니다.

그런데 결국 이는 입력값 유효성 검증과 거의 동일한 성격이라는 점에서 헷갈리기 시작했습니다.
사용자의 잘못된 입력을 막는 책임과,
내부에서 일어나는 버그를 방지하고 싶은 책임이 모델 안에서 섞여있었던 것 같아요. (검증 중복?)

## 개선 방향

`spring-boot-starter-validation` 의존성을 추가합니다.

```java
// 요청 Dto에 검증 로직 추가
public record ReservationRequestDto(
        // @NotEmpty, @NotNull 등으로 기술적으로 유효한가를 판단합니다.
        @NotEmpty(message = "예약자명은 필수입니다.") String name,
        @NotNull(message = "예약 날짜는 필수입니다.") LocalDate date,
        @NotNull(message = "예약 시간은 필수입니다.") LocalTime time
) {

    public Reservation toReservation() {
        // 이미 검증된 값으로만 생성되기 때문에, Reservation의 생성자 내 null 체크는 제거해도 됩니다.
        return new Reservation(name, date, time);
    }
}

// Controller...
@PostMapping
public ReservationResponseDto reserve(@Valid @RequestBody ReservationRequestDto reservationRequestDto) { // @Valid 추가
    Reservation reservation = reservationRepository.add(reservationRequestDto.toReservation());
    return ReservationResponseDto.from(reservation);
}
```

위와 같이 구성되면, 컨트롤러 앞단에서 기술적 형식 검증을 수행할 수 있습니다.
대신 형식 검증에서 발생하는 예외는 `MethodArgumentNotValidException`입니다.
그래서 `RequestExceptionHandler#handleMethodArgumentNotValidException`를 구현하여
형식 검증을 통과하지 못한 경우 `400 Bad Request`와 어떤 필드가 검증을 통과하지 못했는지 반환하도록 했습니다.

## 결론

리뷰 한줄로 꽤 많은 고민을 한 것 같지만..!
결과적으로 기술적 형식 검증은 컨트롤러 앞단에,
비즈니스 규칙 검증은 모델 내에 위치하게 되었습니다.
따라서 새로운 규칙이 추가된다면, 각 유형에 맞는 위치에 추가할 수 있을 것 같습니다! 😄 

[refactor: 입력값의 형식 검증과 비즈니스 규칙 검증 분리](https://github.com/woowacourse/spring-roomescape-admin/pull/265/commits/a0ca7a31eaca9fd429855648ba3e982e6c664fcd)

asebn1 replied:
많은 고민을 하신게 느껴지네요 ㅎㅎ
잘 정리해주셨습니다 :)

asebn1 commented on src/main/java/roomescape/dto/ReservationRequestDto.java:
```diff
@@ -0,0 +1,16 @@
+import java.time.LocalTime;
+import roomescape.model.Reservation;
+
+public record ReservationRequestDto(
```
`Dto`라는 네이밍은 제거해도 될 것 같아요~

asebn1 commented on src/main/java/roomescape/dto/ReservationRequestDto.java:
```diff
@@ -0,0 +1,16 @@
+        LocalTime time
+) {
+
+    public Reservation toReservation() {
+        return new Reservation(name, date, time);
+    }
```
controller에서 사용되는 dto가 도메인을 의존하고 있네요
괜찮을까요?

- https://tecoble.techcourse.co.kr/post/2021-04-25-dto-layer-scope/
- https://umbum.dev/1206/

YehyeokBang replied:
어디선가는 DTO에서 도메인 모델로의 매핑이 필요하다고 판단했고, 다음과 같은 대안을 생각해봤습니다.

컨트롤러에서 직접 도메인 모델 객체 생성 `new Reservation(...);`
→ 컨트롤러가 모델 생성 책임까지 가지게 되어, 역할이 과도하게 확장된다고 느껴 보류했습니다.

전용 매핑 클래스를 만들어 DTO → 모델 변환 담당
→ 매핑이 필요한 도메인 모델 객체마다 클래스가 늘어나 관리하기 어려워질 것 같아서 보류했습니다.

DTO 내부에서 도메인 객체로 변환하는 메서드 포함 (현재 방식)
→ 도메인을 외부 계층에 노출하는 구조이긴 하지만, 컨트롤러가 깔끔하게 유지됩니다.
추가로 매핑 책임을 DTO가 일관되게 갖는다는 점에서 현재는 이 방법을 선택했습니다.

서비스 계층에서 변환 처리
→ 변환 로직이 점차 많아지거나, 계층 분리를 명확히 해야 할 시점에 고려하면 좋을 것 같습니다.

현재는 클래스 수나 계층이 복잡하지 않은 상황이라서
깔끔하고 일관될 수 있게 DTO가 도메인 모델로 변환하는 책임을 가지도록 했습니다.
저와 페어는 `도메인이 DTO를 알게 되는 것만큼은 피해야 한다`는 원칙을 세웠고,
도메인 모델이 DTO에 의존하지 않고, DTO가 도메인 모델을 참조하는 구조는 비교적 괜찮다고 판단했습니다.

### 추가 고민, 의존 방향과 계층 흐름

첨부해주신 글을 읽고, 관련된 다른 글도 찾아봤는데
의존성은 바깥 계층에서 안쪽 계층으로 흐르는 것이 좋다고 느끼는 것 같아요.

```
클라이언트 → DTO → 컨트롤러 → (중간 계층 등) -> 도메인 모델
```

하지만 현재 구조는 아래처럼 보일 수도 있고,

```
클라이언트 → 컨트롤러 → DTO → 도메인 모델 -> (중간 계층 등) # 현재
```

이 과정에서 DTO가 도메인 모델을 생성한다는 점이,
흐름 혼란이나 계층 간 방향성을 해칠 수 있겠다는 생각도 하게 되었습니다.

### 정리 

현재는 간결함, 역할 분리, 의존 방향의 등을 고려하여 실용(?)적인 방법을 선택하게 되었습니다.
이후 여러 모델이 연관되어 변환 로직이 복잡해지거나 많아지는 경우에는
서비스 계층을 도입해 컨트롤러를 더 가볍게 만들고, 변환 로직을 별도로 분리하는 방식을 선택할 수 있을 것 같습니다! 👍 

마지막으로 리뷰의 의도를 잘 이해했는지 또는 제 선택에 관한 제로의 조언이 궁금합니다!
감사합니다! 😄

asebn1 replied:
고민되는 점을 잘 확인하였습니다.
실용성을 중시한 판단은 지금 구조에서는 납득이 가고, 컨트롤러를 가볍게 유지할 수 있다는 장점도 있을 것 같아요.
다만,  말씀하신 것처럼 DTO가 도메인 객체 생성을 직접 담당하는 구조는 계층 간 방향성에 혼동을 줄 수 있을 것 같아요.
프로젝트 규모가 커지고 도메인 모델이 복잡해질수록 DTO가 특정 모델에 강하게 결합되고, 테스트나 재사용에도 제약이 생길 수 있어 저는 초기 개발 단계에서도 가급적 지양하는 편입니다.
상황에 따라 유연하게 대응하려는 방향은 좋다고 생각합니다 👍

YehyeokBang replied:
> 프로젝트 규모가 커지고 도메인 모델이 복잡해질수록 DTO가 특정 모델에 강하게 결합되고, 테스트나 재사용에도 제약이 생길 수 있어 저는 초기 개발 단계에서도 가급적 지양하는 편입니다.

단점은 인지했지만, 장점도 괜찮은 편이라 나중에 (상황에 따라) 필요하면 개선하는 것도 방법이지만,
경험에 의해 미래가 어느 정도 예측되는 부분(DTO <-> 모델 강결합, 재사용 문제 등)의 문제라면,
그 문제 자체가 발생하지 않게 초기에 의식적으로 해결(분리)하는 것도 좋은 방법이 될 것 같네요. 👍 

```
상황에 따라 유연하게 개선하는 접근도 좋은 전략일 수 있지만,
이미 예상 가능한 문제라면 애초에 설계로 예방하는 것도 좋은 전략일 수 있다.
```

를 배운 것 같습니다. 좋은 리뷰 감사합니다! 😄

asebn1 commented on src/main/java/roomescape/model/Reservation.java:
```diff
@@ -0,0 +1,75 @@
+        }
+    }
+
+    public static Reservation toEntity(Integer id, Reservation reservation) {
+        return new Reservation(id, reservation.name, reservation.date, reservation.time);
+    }
```
지금 구조는 도메인과 엔티티가 같은 `Reservation`으로 사용하고있는 것 같아요
그 과정에서 `toEntity`가 있으니 어색하다고 느껴지네요
네이밍을 변경하거나 entity 클래스가 있어야할 것 같아요

YehyeokBang replied:
처음에 리뷰를 받고, 도메인 모델과 엔티티의 역할이 섞여 있기 때문에 `toEntity()` 메서드가 어색하게 보인다고 생각했습니다. 이미 Reservation인데 `toEntity()` 메서드로 다시 (id가 포함된) Reservation을 반환하는 과정이 어색함의 주요 원인이었던 것 같습니다.

그래서 두 단어의 의미를 정의하고, 분리를 시도해봤습니다.

```java
// 비즈니스 유효성을 검증하는 도메인 모델
public final class Reservation {
    
    private final String name;
    private final LocalDate date;
    private final LocalTime time;
    
    // 비즈니스 로직..
    
    public ReservationEntity toEntity(Integer id) {
        // 엔티티 객체로 변환
    }
}

// 각각을 식별할 수 있고, 저장에 유리한 구조를 가지는 엔티티
public final class ReservationEntity {

    private final Integer id;
    private final String name;
    private final LocalDateTime dateTime;
}
```

모델에서는 date와 time을 그대로 가져가면서 각 필드에 유효성을 검증할 수 있습니다. 엔티티는 저장에 유리하도록 dateTime으로 하나의 필드로 관리합니다.

- 도메인 모델은 `문제를 잘 표현하는 것`이 목적
- 반면 엔티티는 `식별과 저장하기 쉬운 구조`가 목적

저장하기 쉬운 구조가 목적이라고는 했지만, `엔티티 == DB 레코드`는 아닌 것 같아요.
식별할 수 있다는 점이 더 중요해보이는..? 참 어려운 것 같습니다.. 😢 

그러나 분리를 해보면서 지금 구조에서 둘을 구분하여 얻을 수 있는 장점이 크지 않다고 느꼈습니다.

- 개념적으로 분리하여 다르게 활용한다는 것 말고는 크게 와닿지 않았습니다.
- 오히려 이미 잘 작동하는데, 매핑하는 단계가 하나 추가된 것 같다는 생각을 하게 되었습니다.
    - 저장된 엔티티를 가져와서 다시 모델로 매핑하는 과정도 포함됩니다.
    - 변환 과정, 각 과정에서 사용되는 것(변수명, 메서드명) 이름 짓기가 애매했습니다.

정리하면,  지금은 `Reservation`의 필드가 간단하고, 저장 구조와 크게 다르지 않기 때문에 분리하지 않아도 될 것 같다고 판단했습니다. 대신, 어색함의 주범이었던 `toEntity()` 메서드명을 `createWithId()`로 변경했습니다.

다만, 언젠가는 모델과 엔티티를 완전히 분리해야 할 상황이 올 수도 있겠다는 생각을 하게 되었습니다.
이 부분을 고려하여, 변경 시 부담을 줄여보고자 했습니다.

- `id`를 받지 않는 생성자를 유지
- 모델을 테스트할 때는 항상 `id` 없이 생성하는 방식으로 사용하도록 변경

이렇게 하면, 나중에 `Reservation`에서 `id` 필드를 제거하고 엔티티를 완전히 분리하더라도,
기존에 작성된 모델 테스트 코드에는 영향을 주지 않도록 대비할 수 있을 것 같습니다. 👍 

[refactor: 혼재된 모델과 엔티티 개념 정립](https://github.com/woowacourse/spring-roomescape-admin/pull/265/commits/16f1a689c7d3417f83a5f4bf7a16954dcbb3c64f)

asebn1 replied:
좋습니다.
다만 현재 `reservation.createWithId(nextId.getAndIncrement());` 형태로 사용하는 것 같은데, 도메인 객체가 새로운 도메인 객체를 만들어 반환하는 것이라 조금 부자연스러운 느낌이 있습니다. (정적 팩토리 메서드같은 네이밍 느낌이네요)
팩토리 메서드로 사용하거나 생성자를 통해 처리하는 방법도 고민해볼 것 같아요.

asebn1 commented on src/test/java/roomescape/MissionStepTest.java:
```diff
@@ -9,12 +17,61 @@
                 .statusCode(200);
     }
 
+    @DisplayName("2단계 - 예약 조회 화면 응답 성공 및 예약 목록 조회 성공")
+    @Test
+    void step2() {
```
DisplayName을 잘 작성해주셨네요~

YehyeokBang replied:
감사합니다! 😄

asebn1 reviewed (changes requested):
벨로 안녕하세요.
방탈출 리뷰어 제로입니다.
잘 작성해주셔서 코드리뷰하기 편했습니다~
몇 가지 코멘트 남겼습니다. 
확인해주세요!

YehyeokBang committed:
refactor: 혼재된 모델과 엔티티 개념 정립

- id가 추가된다고 모델에서 엔티티가 되는 것처럼 느껴지던 메서드명을 수정합니다.
- 현재는 두 개념을 두 클래스로 명확히 분리하지 않습니다.

YehyeokBang committed:
refactor: 입력값의 형식 검증과 비즈니스 규칙 검증 분리

- 기술적인 형식 검증은 컨트롤러 앞단에서 진행합니다.
- 입력값의 비즈니스 규칙 검증은 도메인 모델에서 진행합니다.
- 검증 관련 의존성을 추가합니다.

YehyeokBang committed:
refactor: 예외 클래스명 변경

- 기존 추상적인 도메인 예외라는 이름 대신 각 모델에 맞게 이름을 변경합니다.

YehyeokBang reviewed (commented):
감사합니다!
이번에도 잘 부탁드려요! 😄

asebn1 reviewed (approved):
벨로 안녕하세요
리뷰 잘 마무리해주신것 확인하였습니다
이번 단계는 이만 머지하겠습니다
다음단계도 화이팅!
